### PR TITLE
gmp: bump + enable building static libraries on MSVC

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **gmp/all**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/gmp/config.yml [hunk 1]
```diff
@@ -1,5 +1,7 @@
 versions:
-  "6.1.2":
-    folder: all
   "6.2.0":
     folder: all
+  "6.2.1":
+    folder: all
```
Comment (madebr) [inline recipes/gmp/config.yml:5]:
```suggestion
  "6.2.1":
    folder: all
  "6.2.0":
    folder: all
```

File: recipes/gmp/all/conanfile.py [hunk 2]
```diff
@@ -36,46 +42,79 @@ def package_id(self):
...
+            "--srcdir={}".format(os.path.join(self.source_folder, self._source_subfolder).replace("\\", "/")),
+        ]
+        if self.settings.compiler == "Visual Studio":
+            configure_args.extend([
+                "ac_cv_c_restrict=restrict",
+                "gmp_cv_asm_label_suffix=:",
+                "lt_cv_sys_global_symbol_pipe=cat",  # added to get further in shared MSVC build, but it gets stuck later
+            ])
+            self._autotools.flags.append("-FS")
+            # self._autotools.cxx_flags.append("-EHsc")
```
Comment (madebr) [inline recipes/gmp/all/conanfile.py:79]:
```suggestion
```

File: recipes/gmp/all/conanfile.py [hunk 3]
```diff
@@ -36,46 +42,79 @@ def package_id(self):
...
-        with tools.chdir(self._source_subfolder):
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            tools.patch(**patch)
+        with self._build_context():
             autotools = self._configure_autotools()
             autotools.make()
-        # INFO: According to the gmp readme file, make check should not be omitted, but it causes timeouts on the CI server.
-        if self.options.run_checks:
-            autotools.make(args=["check"])
+            # INFO: According to the gmp readme file, make check should not be omitted, but it causes timeouts on the CI server.
```
Comment (prince-chrismc) [inline recipes/gmp/all/conanfile.py:110]: 🏴

File: recipes/gmp/config.yml (hunk 1 above)
Comment (prince-chrismc) [reply recipes/gmp/config.yml:5]: Is it work adding `6.2.0` ... maybe justy keep the point release?

File: recipes/gmp/all/conanfile.py (hunk 3 above)
Comment (madebr) [reply recipes/gmp/all/conanfile.py:110]: I think this is correct.
The tests are run in the build context.
The checks are also disabled by default, so there will be no timeouts.

File: recipes/gmp/config.yml (hunk 1 above)
Comment (madebr) [reply recipes/gmp/config.yml:5]: I add 6.2.1 in this pr.
This comment was just about ordering the versions in descending order.

**Review by prince-chrismc — APPROVED** [review]

**Comment by madebr** [comment]
@SpaceIm @uilianries @gocarlos @Croydon 
Please review. Thanks!

File: recipes/gmp/all/conanfile.py [hunk 4]
```diff
@@ -36,46 +42,79 @@ def package_id(self):
         del self.info.options.run_checks  # run_checks doesn't affect package's ID
 
     def build_requirements(self):
-        if tools.os_info.is_windows and self.settings.compiler != "Visual Studio" and \
-           "CONAN_BASH_PATH" not in os.environ and tools.os_info.detect_windows_subsystem() != "msys2":
+        if tools.os_info.is_windows and not tools.get_env("CONAN_BASH_PATH"):
             self.build_requires("msys2/20200517")
+        if self.settings.compiler == "Visual Studio":
+            self.build_requires("yasm/1.3.0")
```
Comment (SpaceIm) [inline recipes/gmp/all/conanfile.py:48]: Why only for Visual Studio? Is it not also required for all platforms if `disable_assembly=False`?

Comment (madebr) [reply recipes/gmp/all/conanfile.py:48]: Other platforms can just use the gnu assembler.

Comment (madebr) [reply recipes/gmp/all/conanfile.py:48]: By using yasm, this also means that for the moment only x86 and x86_64 architectures are supported on MSVC.
Maybe it is possible to combine the GNU assembler (from binutils) with MSVC, but that's something I did not test.

Comment (SpaceIm) [reply recipes/gmp/all/conanfile.py:48]: Ok, so LGTM

**Review by SpaceIm — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**